pkg/osquery/tables/firefox_preferences/table.go

```diff
@@ -0,0 +1,85 @@
+package firefox_preferences
+
+import (
+	"bufio"
+	"context"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	"github.com/kolide/launcher/pkg/dataflatten"
+	"github.com/kolide/launcher/pkg/osquery/tables/dataflattentable"
+	"github.com/kolide/launcher/pkg/osquery/tables/tablehelpers"
+	"github.com/osquery/osquery-go"
+	"github.com/osquery/osquery-go/plugin/table"
+	"github.com/pkg/errors"
+)
+
+type Table struct {
+	name   string
+	logger log.Logger
+}
+
+const tableName = "kolide_firefox_preferences"
+
+func TablePlugin(_ *osquery.ExtensionManagerClient, logger log.Logger) *table.Plugin {
+	columns := dataflattentable.Columns(
+		table.TextColumn("path"),
+	)
+
+	t := &Table{
+		name:   tableName,
+		logger: logger,
+	}
+
+	return table.NewPlugin(t.name, columns, t.generate)
+}
+
+func (t *Table) generate(ctx context.Context, queryContext table.QueryContext) ([]map[string]string, error) {
+	return generateData(queryContext, t.logger)
+}
+
+func generateData(queryContext table.QueryContext, logger log.Logger) ([]map[string]string, error) {
+	paths := tablehelpers.GetConstraints(queryContext, "path")
+
+	if len(paths) != 1 {
+		return nil, errors.Errorf("The %s table requires that you specify a constraint for path", tableName)
+	}
+
+	file, err := os.Open(paths[0])
+	if err != nil {
+		// TODO: Investigate what error message looks like. Add filepath possibly
+		return nil, err
+	}
+
+	scanner := bufio.NewScanner(file)
+
+	rowData := map[string]string{"path": paths[0]}
+	rawKeyVals := make(map[string]interface{})
+
+	re := regexp.MustCompile(`user_pref\((.*)\)`)
+	for scanner.Scan() {
+		line := scanner.Text()
+		match := re.FindStringSubmatch(line)
+
+		if len(match) <= 1 {
+			continue
+		}
+
+		parts := strings.Split(match[1], ", ")
+		rawKeyVals[parts[0]] = parts[1]
```

Comment:
I don't think `parts[1]` is guaranteed? If not this will panic.

Reply:
It should be guaranteed once it reaches this point. If the regex matches, then the string returned will always be a key/val pair separated by a comma.

Reply:
I don't understand how `^user_pref\((.*)\)` guarantees that. Yes, if the file conforms to your expectations. But it may not.

Reply:
That makes sense. I was thinking that if the file doesn't conform to what we expect, then we have larger problems. I'll check to make sure `len(parts) == 2` before adding the key/val.

Reply:
This comment should be paired with the above one about split. I think you have several options.

If I understand correctly, you're assuming a structure:
* lines match `user_pref("KEY", VALUE);`
* `KEY` does not include quotes
* `VALUE` might be arbitrarily complex and gnarly
* All other parts of that are assumed as fixed. (eg, ends with `);`

I think your first option is to use a more clever regex. You can extract the KEY and the VALUE directly. With this code, this is the direction I'd go.

I think the second option is to limit the split. In go, that'd be `SplitN`, in ruby it's the 3 argument split. I would bias away from this one.

Last choice is to use (or write) a parsing library

Reply:
> if the file doesn't conform to what we expect, then we have larger problems

Maybe. It might have a format we haven't seen, or it might be a broken/truncated/corrupt file, either accidentally or intentionally. While we might not be able to extract data, we should never panic or throw errors.